src/SFA.DAS.Authorization/Cache/AuthorizationCacheService.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace SFA.DAS.Authorization.Cache
+{
+    public class AuthorizationCacheService : IAuthorizationCacheService
+    {
+        private readonly IMemoryCache _memoryCache;
+        private readonly Dictionary<Type, IAuthorizationHandlerCacheConfig> _cacheKeyProviders;
+
+        public AuthorizationCacheService(IMemoryCache memoryCache, IAuthorizationHandlerCacheConfig[] config)
+        {
+            _memoryCache = memoryCache ?? throw new ArgumentNullException(nameof(memoryCache));
+
+            if (config == null) throw new ArgumentNullException(nameof(config));
+
+            _cacheKeyProviders = config
+                .SelectMany(cp => cp.SupportsHandlerTypes
+                                    .Select(sht => new {
+                                        KeyProvider = cp,
+                                        HandlerType = sht
+                                    }))
+                .ToDictionary(cp => cp.HandlerType, kp => kp.KeyProvider);
+        }
+
+        public Task<AuthorizationResult> GetOrAdd(
+            IAuthorizationHandler handler, 
+            IReadOnlyCollection<string> options,
+            IAuthorizationContext authorizationContext)
+        {
+            var cacheDetails = GetCacheDetailsForAuthorizationHandlerType(handler, options, authorizationContext);
```

Comment:
I think it's worth allowing the wrapped authorization handler to run first before building a cache key as there are a few scenarios where it will actually throw an exception (e.g. unrecognized options, missing/invalid authorization context keys, etc).

Reply:
Not sure I understand this point. Doesn't that mean we'll always be calling the wrapped handler, thus removing any benefit in the cache?